enhancements/hypershift/support-disabling-image-registry-via-capability.md

```diff
@@ -0,0 +1,573 @@
+    // +listType=atomic
+    // +optional
+    DisabledCapabilities []OptionalCapability
+}
+```
+
+With the new field in place, the `cvoBootrapScript` needs to create the
+`ClusterVersion` resource with a specific set of enabled capabilities, minus
+the ones specified in `DisabledCapabilities`.
+This function currently doesn't have access to the `HostedClusterSpec`.
+`cvo.ReconcileDeployment` will need to change to allow it access to the set of
+disabled capabilities specified by the customer.
+
+TODO: all the reconcile functions in `hostedcontrolplane_controller.go` take a
+`HostedControlPlane` object, not a `HostedCluster`. My assumption is that the
+`HosterControlPlane` gets created from the `HostedCluster`. If this is correct,
+where does it happen?
+
+### Topology Considerations
+
+#### Hypershift / Hosted Control Planes
+
+Hypershift is the target topology of this enhancement.
+
+#### Standalone Clusters
+
+Standalone clusters already support capabilities - the changes in this
+enhancement do not affect this topology.
+
+#### Single-node Deployments or MicroShift
+
+The changes in this enhancement target Hypershift specific API and CLI.
+Single-node deployments and Microshift are unnaffected.
+
+### Implementation Details/Notes/Constraints
+
+#### Evolving the Hypershift capabilities API
+
+With a capabilities framework in place, it's expected that more capabilities
+become available for disabling at install time. Currently in OpenShift
+Standalone all optional capabilities are enabled by default. In the future, that
+might change and some optional capabilities might come disabled by default.
+
+In that case, cluster service consumers might need a way to decide which
+of the disabled optional capabilities they want to re-enable at cluster install
+time. In this case, a `--enabled-capability` flag may be added to the Hypershift
+CLI, and an equivalent `EnabledCapabilities` added to the `HostedCluster`
+resource.
+
+When extending the list of supported capabilities in Hypershift, it is crucial
+to use existing capability names from the CVO API. Standalone cluster
+capability names should exactly match the ones supported by Hypershift, although
+the set of supported capabilities is expected to differ.
+
+#### Re-enabling components
+
+When the cluster service consumer re-enables the `ImageRegistry` capability on a
+running cluster, the CVO will provision the resources for the cluster-image-registry-operator,
+resulting in new pods in the management cluster. Additionally, the pull secrets
+controller (currently located in the openshift-controller-manager) will also be
+enabled by Hypershift, causing a pod redeployment.
+
+### Risks and Mitigations
+
+#### Lack of implicitly enabled capabilities
+
+During hosted control plane upgrades, the CVO is currently unaware it's being
+upgraded. This leads to the CVO being unable to calculate what capabilities
+should be implicitly enabled for clusters during upgrades.
```

Comment:
is this simply a matter of including the information about capabilities in a manifest in the payload for consumption by the CVO?  We do something similar for featuregates today.

Reply:
No, implicit capabilities depend on [the past state of the cluster][1] compared with the current target release.  We already have all the information we need about the current target release.  And we can't bake the past state of the cluster (which is cluster-specific) into release images.

[1]: https://github.com/openshift/enhancements/blob/master/enhancements/installer/component-selection.md#updates

Reply:
Summarizing yesterday's chat here: while we agree that implicitly enabled capabilities is something that needs to be solved in Hypershift, specially when existing capabilities stop being enabled by default, it's not something we're going to cover in this EP, as the changes proposed here do not introduce new problems or exacerbate the existing issues.

I'll update the EP to reflect this.